aspnet/fundamentals/dependency-injection.rst

```diff
@@ -1,10 +1,277 @@
+
+Replacing the default services container
+----------------------------------------
+
+The built-in services container is mean to serve the basic needs of the framework and most consumer applications built on it. However, developers who wish to replace the built-in container with their preferred container can easily do so. The ``ConfigureServices`` method typically returns ``void``, but if its signature is changed to return ``IServiceProvider``, a different container can be configured and returned. There are `many IOC containers available for .NET <https://github.com/danielpalme/IocPerformance/blob/master/README.md>`_. This article will attempt to add links to DNX implementations of containers as they become available. In this example, the `StructureMap.Dnx <https://github.com/structuremap/structuremap.dnx>`_ package is used.
+
+First, add the appropriate container to the dependencies property in project.json:
+
+.. code-block:: javascript
+
+	"dependencies" : {
+		"StructureMap.Dnx": "0.4.0-alpha"
+	},
+
+.. note:: Currently StructureMap does not support .NET Core.
+
+Next, configure the container in ``ConfigureServices`` and return an ``IServiceProvider``:
+
+.. code-block:: c#
+
+	public IServiceProvider ConfigureServices(IServiceCollection services)
+	{
+		services.AddMvc();
+		// add other services
+		
+		// Add StructureMap
+		var container = new StructureMap.Container();
+		container.Populate(services);
+		return container.GetInstance<IServiceProvider>();
+	}
+
+.. note:: If you add StructureMap to the sample project, you will need to add a ``[DefaultConstructor]`` attribute to the default constructor for ``Operation``.
+
+.. list-table:: ASP.NET 5 / DNX Containers (in alphabetical order)
+   :header-rows: 1
+
+   * - Package (Nuget)
+     - Setup Instructions
+   * - `Autofac.Dnx <http://www.nuget.org/packages/Autofac.Dnx/>`_
+     - http://autofac.org/
+   * - `StructureMap.Dnx <http://www.nuget.org/packages/StructureMap.Dnx>`_
+     - `Getting Started <https://github.com/structuremap/structuremap.dnx/blob/master/README.md>`_
+
+Recommendations
+---------------
+
+When working with dependency injection, keep the following recommendations in mind:
+
+- DI is for objects that have complex dependencies. Controllers, services, adapters, and repositories are all examples of objects that might be added to DI.
+- Avoid storing data and configuration directly in DI. For example, a user's shopping cart shouldn't typically be added to the services container. Configuration should use the :doc:`options`.
+- Avoid creating "data holder" objects. These are objects that only exist to allow access to some other object. It's better to request the actual item needed via DI, is possible.
```

Comment:
Discussed this exact thing with @rynowak, who proposed this recommendation. Added a note to call out the framework's use of this practice, and emphasizing that doing this should be rare in application code.